apps/linear_algebra/src/blas_l1_generators.cpp

```diff
@@ -0,0 +1,219 @@
+#include <vector>
+#include "Halide.h"
+
+using namespace Halide;
+
+namespace {
+
+// Generator class for BLAS axpy operations.
+template<class T>
+class AXPYGenerator :
+        public Generator<AXPYGenerator<T>> {
+  public:
+    typedef Generator<AXPYGenerator<T>> Base;
+    using Base::target;
+    using Base::get_target;
+    using Base::natural_vector_size;
+
+    GeneratorParam<bool> assertions_enabled_ = {"assertions_enabled", false};
+    GeneratorParam<bool> use_fma_ = {"use_fma", false};
+    GeneratorParam<bool> vectorize_ = {"vectorize", true};
+    GeneratorParam<bool> parallel_ = {"parallel", true};
+    GeneratorParam<int>  block_size_ = {"block_size", 1024};
+    GeneratorParam<bool> scale_x_ = {"scale_x", true};
+    GeneratorParam<bool> add_to_y_ = {"add_to_y", true};
+
+    // Standard ordering of parameters in AXPY functions.
+    Param<T>   a_ = {"a", 1.0};
+    ImageParam x_ = {type_of<T>(), 1, "x"};
+    ImageParam y_ = {type_of<T>(), 1, "y"};
+
+    void SetupTarget() {
+        if (!assertions_enabled_) {
+            target.set(get_target()
+                       .with_feature(Target::NoAsserts)
+                       .with_feature(Target::NoBoundsQuery));
+        }
+
+        if (use_fma_) {
+            target.set(get_target().with_feature(Target::FMA));
+        }
+    }
+
+    void Schedule(Func result, Expr width) {
+        Var i("i"), o("o");
+    }
+
+    Func build() {
+        SetupTarget();
+
+        const int vec_size = vectorize_? natural_vector_size(type_of<T>()): 1;
+
+        bool scale_x = scale_x_;
+        bool add_to_y = add_to_y_;
+
+        Var  i("i");
+        Func result("result");
+        if (scale_x && add_to_y) {
+            result(i) = a_ * x_(i) + y_(i);
+        } else if (scale_x) {
+            result(i) = a_ * x_(i);
+        } else {
+            result(i) = x_(i);
+        }
+
+        if (vectorize_) {
+            Var ii("ii");
+            result
+                .specialize(x_.width() >= vec_size).vectorize(i, vec_size)
+                .specialize(x_.width() >= 4 * vec_size).unroll(i, 4);
+        }
+
+        if (parallel_) {
+            Var ii("ii");
+            Expr factor = block_size_ / vec_size;
+            result.specialize(x_.width() >= 4 * block_size_).split(i, i, ii, factor).parallel(i);
```

Comment:
True, the parallel schedules in the L1 generators are pretty messed up because I have not been able to get parallelism to give use any performance improvement. So I probably need to do a cleanup of these chunks in general.